File: lib/ts/src/internal/binary/Pack.ts
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

import { packList } from '../../internal/binary/PackList.js';

export function pack(value: any): any {
    if (Array.isArray(value)) {
        return packList(value);
    } else if (value instanceof Map) {
        const newMap: Map<any, any> = new Map();

        for (const [key, val] of value.entries()) {
            newMap.set(key, pack(val));
        }

        return newMap;
    } else {
        return value;
    }
}

File: lib/go/internal/binary/Pack.go
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package binary

// Pack attempts to pack nested maps/lists into the compact binary representation used by Telepact.
func Pack(value any) (any, error) {
	switch typed := value.(type) {
	case []any:
		return PackList(typed)
	case map[string]any:
		result := make(map[string]any, len(typed))
		for key, val := range typed {
			packed, err := Pack(val)
			if err != nil {
				return nil, err
			}
			result[key] = packed
		}
		return result, nil
	case map[int]any:
		result := make(map[int]any, len(typed))
		for key, val := range typed {
			packed, err := Pack(val)
			if err != nil {
				return nil, err
			}
			result[key] = packed
		}
		return result, nil
	case map[any]any:
		result := make(map[any]any, len(typed))
		for key, val := range typed {
			packed, err := Pack(val)
			if err != nil {
				return nil, err
			}
			result[key] = packed
		}
		return result, nil
	default:
		return value, nil
	}
}
